###PHYLOSEQ SCRIPT


install.packages("BiocManager", repos = "http://cran.r-project.org")

install.packages("devtools")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager") 
BiocManager::install("phyloseq",force = TRUE)

#if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager") 
BiocManager::install("microbiome",force = TRUE)
BiocManager::install("edgeR",force = TRUE)

options(repos = c(CRAN = "https://cran.rstudio.com"))

install.packages("readxl", repos = "http://cran.r-project.org")
install.packages("devtools", repos = "http://cran.r-project.org")


install.packages("readxl", repos = "http://cran.r-project.org")
install.packages("roxygen2", repos = "http://cran.r-project.org")
install.packages("limma", repos = "http://cran.r-project.org")

install.packages("ggpubr", repos = "http://cran.r-project.org")

library(readxl)
#install.packages("edgeR")
source("fcns/config.r")
#setwd("external/altindis/celiac_disease_16sn/phyloseq"))
#install.packages("Rtools")
#install.packages("phyloseq")
#install.packages("phyloseq", dependencies = TRUE)

library(remotes)
remotes::install_github(repo="jdreyf/ezlimma", build_opts = c("--no-resave-data", "--no-manual"), force = TRUE) 
remotes::install_github(repo="jdreyf/ezlimmaplot", build_opts = c("--no-resave-data", "--no-manual"), force = TRUE) 
remotes::install_github(repo="jdreyf/Hitman", build_opts = c("--no-resave-data", "--no-manual"), force = TRUE) 
remotes::install_github(repo="jdreyf/jdcbioinfo", build_opts = c("--no-resave-data", "--no-manual"), force = TRUE)
devtools::install_github("rstudio/rmarkdown")
source("fcns/config.r")
#setwd("external/altindis/celiac_disease_16sn/phyloseq")
library(phyloseq)
library(vegan)
library(microbiome)
library(ggplot2)
library(ggpubr)

## Purpose
#Overview of 16S DNA-seq of gingiva samples by Phyloseq [1].  
#The results are in folder [phyloseq](../phyloseq/).  

## Data
#OTU table [seqtab_nochim.rds](../dada2/seqtab_nochim.rds) and taxonomy annotation [annot_add_species.rds](../dada2/annot_add_species.rds) were generated by using R package *dada2* [2]. We merge the two technical replicates for each Sample.
#Metadata [CeD new label (2).csv](../data/CeD new label (2).csv) and [Celiakibarn 190906.csv](../data/Celiakibarn 190906.csv).  

#```{r parse}
# meta data
md_old <- read.csv("data/CeD new label (2).csv")
md_age <- read.csv("data/Celiakibarn 190906.csv")
md_old <- merge(md_old, md_age, all = TRUE)
rm(md_age)

##DD removed anything but age 1 samples
md <- subset(md_old, Age==1.0)
# md$Disease.status[!is.na(md$Age.at.diagnosis)] <- ifelse(md$Age.at.diagnosis[!is.na(md$Age.at.diagnosis)] <= 5, "A0", md$Disease.status[!is.na(md$Age.at.diagnosis)])

md <- md[order(md$New.Sample.ID), ]
md <- md[rep(1:nrow(md), each = 3),]
md$New.Sample.ID <- paste0("S", md$New.Sample.ID, c("", ".neg", ".pos"))
md$Sort <- rep(c("Presort", "IGneg", "IGpos"), time = nrow(md)/3)
md$Disease.status <- sapply(md$Disease.status, FUN = switch, A = "Celiac", B = "Control")
md$ABIS.no <- paste0("ABIS", md$ABIS.no)

md$Disease.status <- factor(md$Disease.status, levels = c("Control", "Celiac"))
md$Sorted <- md$Sort <- factor(md$Sort, levels = c("Presort", "IGneg", "IGpos"))
levels(md$Sorted) <- list(Presort = "Presort", Postsort = c("IGneg", "IGpos"))
md <- md[order(md$Sort, md$Disease.status, md$Age, md$ABIS.no), ]
md$Group <- paste(md$Sort, md$Disease.status, md$Age, sep = "_")
md$Group <- factor(md$Group, levels = unique(md$Group))
md$Age <- factor(paste0(md$Age, "year"))
md <- data.frame(md, row.names = "New.Sample.ID")

## otu
seqtab.nochim <- readRDS("dada2/seqtab_nochim.rds")
dim(seqtab.nochim)
grp <- gsub("\\.[1-2]$", "", gsub("\\.[1-2]\\.", ".", rownames(seqtab.nochim)))
seqtab.nochim <- rowsum(seqtab.nochim, group = grp)
rm(grp)
dim(seqtab.nochim)
nms <- intersect(rownames(md), rownames(seqtab.nochim))
seqtab.nochim <- seqtab.nochim[nms, ]
md <- md[nms, ]

# taxa
taxa <- readRDS("dada2/annot_add_species.rds")
stopifnot(colnames(seqtab.nochim) == rownames(taxa))

#make Phyloseq Object
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), sample_data(md[rownames(seqtab.nochim), ]), tax_table(taxa))
#```

## Filter and transformation
#```{r filt}
#Take a look at the read counts across samples
plot(sample_sums(ps))

# filt samp
# ps <-  prune_samples(sample_sums(ps) >= 5000, ps)

# normalize samp
nps <- transform_sample_counts(ps, function(x) x/sum(x))

# filt otu
keep_otu <- genefilter_sample(ps, flist = filterfun_sample(function(x) x >= 5), A = 1)
nps <- prune_taxa(keep_otu, nps)

nps1 <-  transform_sample_counts(nps, function(x) x * 1e6)
#```

#We keep OTUs with at least in 5 counts in 1 sample. We then transform the read counts to proportions (relative abundance) i.e. read counts divided by total counts of each sample.  

## Plot Ordination
#```{r ord}
# NMDS
set.seed(123456)
ord <- ordinate(nps1, method = "NMDS", distance  = "bray", maxit = 1000, sratmax = 25,  sfgrmin = 1e-10)

pdf("nmds_plot_1n.pdf", 9, 7.5)
ggp <- plot_ordination(nps1, ordination = ord, type = "samples", color = "Disease.status") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sort ~ Age)+scale_color_manual(values=c( "#00BFC4","#F8766D"))+ stat_ellipse()
plot(ggp)
dev.off()

  #scale_color_manual(values=c( "green","red"))

pdf("nmds_plot_2n.pdf", 9, 5)
ggp <- plot_ordination(nps1, ordination = ord, type = "samples", color = "Disease.status", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sorted ~ Age)+scale_color_manual(values=c( "#00BFC4","#F8766D"))+ stat_ellipse()
plot(ggp)
dev.off()

# PCA
ord <- ordinate(nps1, method = "RDA", distance  = "euclidean")

pdf("pca_plot_1n.pdf", 9, 7.5)
ggp <- plot_ordination(nps1, ordination = ord, type = "samples", color = "Disease.status") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sort ~ Age) +scale_color_manual(values=c( "#00BFC4","#F8766D"))+ stat_ellipse()
plot(ggp)
dev.off()


pdf("pca_plot_2n.pdf", 9, 5)
ggp <- plot_ordination(nps1, ordination = ord, type = "samples", color = "Disease.status", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sorted ~ Age)+scale_color_manual(values=c( "#00BFC4","#F8766D"))+ stat_ellipse()
plot(ggp)
dev.off()

#```

#To get an overview of sample similarity/dissimilarity, we perform Non-metric MultiDimenstional Scaling (NMDS) analysis and Principle Components Analysis (PCA). The NMDS plots are at [nmds_plot_1.pdf](./nmds_plot_1.pdf), [nmds_plot_2.pdf](./nmds_plot_2.pdf), and the PCA plots are at [pca_plot_1.pdf](./pca_plot_1.pdf), [pca_plot_2.pdf](./pca_plot_2.pdf).  

#```{r}
nps1.sorted <- subset_samples(nps1, Sorted == "Postsort") 
ord.sorted <- ordinate(nps1.sorted, method = "RDA", distance  = "euclidean")

pdf("pca_plot_3n.pdf", 6.5, 3)
ggp <- plot_ordination(nps1.sorted, ordination = ord.sorted, type = "samples", color = "Age", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid( ~ Disease.status) +scale_color_manual(values=c( "#00BFC4","#F8766D"))+ stat_ellipse()
plot(ggp)
dev.off()

pdf("pca_plot_4n.pdf", 5.5, 4)
ggp <- plot_ordination(nps1.sorted, ordination = ord.sorted, type = "samples", color = "Group", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2)+ stat_ellipse()
plot(ggp)
dev.off()
#```

#```{r, adonis}
set.seed(100)
permanova <- list()
for (s in unique(sample_data(nps1)$Sort)) {
  for (a in unique(sample_data(nps1)$Age)) {
    nps1.sub <- subset_samples(nps1, Sort == s & Age  == a)
    aov.tab <- adonis2(otu_table(nps1.sub) ~ Disease.status, data = data.frame(sample_data(nps1.sub)), perm = 1e4-1, method = "bray")$aov.tab
    permanova[[paste(s,a)]] <- c(Sort = s, Age = a, disease.status.pval = aov.tab["Disease.status", "Pr(>F)"])
  }
}
permanova <- Reduce(rbind, permanova)
write.csv(permanova, "bray_curtis_distance_permanova_stats.csv", row.names = FALSE)
#```

## Diversity estimates

#```{r div}
## alpha diversities
mss <- c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher")

pdf("alpha_diversities_plot_1.pdf", 9, 7.5)
for(ms in mss) {
  ggp <- plot_richness(ps, x = "Disease.status", measures = ms) + geom_boxplot(aes(fill = Disease.status))
  ggp <- ggp + theme(legend.position = "none")+scale_fill_manual(values=c( "#00BFC4","#F8766D"))+ theme_bw() + ggtitle(paste("Measure:", ms)) + facet_grid(Sort ~ Age)
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test")
  plot(ggp)
}
dev.off()

pdf("alpha_diversities_plot_2.pdf", 9, 7.5)
for(ms in mss) {
  ggp <- plot_richness(ps, x = "Age", measures = ms) + geom_boxplot(aes(fill = Age)) + theme_bw()
  ggp <- ggp + theme(legend.position = "none")+scale_fill_manual(values=c( "#00BFC4","#F8766D")) + ggtitle(paste("Measure:", ms)) + facet_grid(Sort ~ Disease.status)
  #+stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

## anova
alpha_div <- estimate_richness(ps, split = TRUE, measures = mss)
alpha_div <- data.frame(alpha_div, md[rownames(alpha_div), ])


#Last addition
alpha_div <- estimate_richness(ps, split = TRUE, measures = mss)
alpha_div <- data.frame(alpha_div, md[rownames(alpha_div), ])

anova_1 <- list()

anova_tmp <- matrix(NA, nrow = 3, ncol = 2)
dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))


nrow_anova_tmp <- length(levels(md$Sort))
ncol_anova_tmp <- length(levels(md$Age))
anova_tmp <- matrix(NA, nrow = nrow_anova_tmp, ncol = ncol_anova_tmp)
dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))

anova_1 <- list()
for(ms in mss) {
  # Use the dynamically calculated dimensions
  anova_tmp <- matrix(NA, nrow = nrow_anova_tmp, ncol = ncol_anova_tmp)
  dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))
  
  for(age in levels(md$Age)){
    for(sort in levels(md$Sort)){
      dat_tmp <- alpha_div[md$Age == age & md$Sort == sort, ]
      aov_res <- aov(as.formula(paste(ms," ~ Disease.status")), data = dat_tmp)
      anova_tmp[sort, age] <- summary(aov_res)[[1]]["Disease.status", "Pr(>F)"]
    }
  }
  
  anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
  colnames(anova_tmp)[1] <- ""
  anova_1[[ms]] <- anova_tmp
}

writexl::write_xlsx(anova_1, "alpha_diversities_plot_1_anova_pval.xlsx")


# anova_2 <- list()
# for(ms in mss) {
#   anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
#   dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Disease.status))
#   for(sort in levels(md$Sort)){
#     for(status in levels(md$Disease.status)){
#       dat_tmp <- alpha_div[md$Disease.status == status & md$Sort == sort, ]
#       aov_res <- aov(as.formula(paste(ms," ~ Age")), data = dat_tmp)
#       anova_tmp[sort, status] <- summary(aov_res)[[1]]["Age", "Pr(>F)"]
#     }
#   }
#   anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
#   colnames(anova_tmp)[1] <- ""
#   anova_2[[ms]] <- anova_tmp
# }
# writexl::write_xlsx(anova_2, "alpha_diversities_plot_2_anova_pval.xlsx")

## beta diversities use nps
dis <- distance(nps, method = "bray")
mod <- betadisper(dis, group = sample_data(nps)$Group)
stopifnot(attr(dis, "Labels") == rownames(sample_data(nps)))
stopifnot(attr(mod$distances, "names") == rownames(sample_data(nps)))
dat2p <- cbind(Distance = mod$distances, sample_data(nps))

pdf("beta_diversities_plot_1.pdf", 9, 7.5)
ggp <- ggplot(data = dat2p, aes(x = Disease.status, y = Distance)) + geom_boxplot(aes(fill = Disease.status)) + theme_bw()
ggp <- ggp + theme(legend.position = "none")
ggp <- ggp + labs(y = "Distance to centroid", title = "Bray-Curtis distance") + facet_grid(Sort ~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
#stat_compare_means(label.x=2,label.y=0.65, method = "t.test")
plot(ggp)
dev.off()

pdf("beta_diversities_plot_2.pdf", 9, 7.5)
ggp <- ggplot(data = dat2p, aes(x = Age, y = Distance)) + geom_boxplot(aes(fill = Age)) + theme_bw()
ggp <- ggp + theme(legend.position = "none")
ggp <- ggp + labs(y = "Distance to centroid", title = "Bray-Curtis distance") + facet_grid(Sort ~ Disease.status)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
#stat_compare_means(label.x=2,label.y=0.65, method = "t.test")
plot(ggp)
dev.off()

anova_1 <- list()
for(dis in "Distance") {
  anova_tmp <- matrix(NA, nrow = 3, ncol = 2)
  dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))
  for(age in levels(md$Age)){
    for(sort in levels(md$Sort)){
      dat_tmp <- dat2p[md$Age == age & md$Sort == sort, ]
      aov_res <- aov(as.formula(paste(dis," ~ Disease.status")), data = dat_tmp)
      anova_tmp[sort, age] <- summary(aov_res)[[1]]["Disease.status", "Pr(>F)"]
    }
  }
  anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
  colnames(anova_tmp)[1] <- ""
  anova_1[[dis]] <- anova_tmp
}
writexl::write_xlsx(anova_1, "beta_diversities_plot_1_anova_pval.xlsx")

#new addition
nrow_anova_tmp <- length(levels(md$Sort))
ncol_anova_tmp <- length(levels(md$Age))

anova_1 <- list()
for(dis in "Distance") {
  anova_tmp <- matrix(NA, nrow = nrow_anova_tmp, ncol = ncol_anova_tmp)
  dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))
  
  for(age in levels(md$Age)){
    for(sort in levels(md$Sort)){
      dat_tmp <- dat2p[md$Age == age & md$Sort == sort, ]
      aov_res <- aov(as.formula(paste(dis," ~ Disease.status")), data = dat_tmp)
      anova_tmp[sort, age] <- summary(aov_res)[[1]]["Disease.status", "Pr(>F)"]
    }
  }
  
  anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
  colnames(anova_tmp)[1] <- ""
  anova_1[[dis]] <- anova_tmp
}

writexl::write_xlsx(anova_1, "beta_diversities_plot_1_anova_pval.xlsx")



# anova_2 <- list()
# for(dis in "Distance") {
#   anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
#   dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Disease.status))
#   for(sort in levels(md$Sort)){
#     for(status in levels(md$Disease.status)){
#       dat_tmp <- dat2p[md$Disease.status == status & md$Sort == sort, ]
#       aov_res <- aov(as.formula(paste(dis," ~ Age")), data = dat_tmp)
#       anova_tmp[sort, status] <- summary(aov_res)[[1]]["Age", "Pr(>F)"]
#     }
#   }
#   anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
#   colnames(anova_tmp)[1] <- ""
#   anova_2[[dis]] <- anova_tmp
# }
# writexl::write_xlsx(anova_2, "beta_diversities_plot_2_anova_pval.xlsx")
#```

#First, we estimate the alpha diversities of the different disease status at different age and sorting. We do this by using different measures, namely "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", and "Fisher".  The plots are at [alpha_diversities_plot_1.pdf](./alpha_diversities_plot_1.pdf) and [alpha_diversities_plot_2.pdf](./alpha_diversities_plot_2.pdf). 

#The corresponding ANOVA F-test p-values are at Excel sheets [alpha_diversities_plot_1_anova_pval.xlsx](./alpha_diversities_plot_1_anova_pval.xlsx) and [alpha_diversities_plot_2_anova_pval.xlsx](./alpha_diversities_plot_2_anova_pval.xlsx). For example, presorted data from kids at 1year, we test Observed alpha diversity vs disease status with an [F-test](https://en.wikipedia.org/wiki/F-test). This would be a correlation if disease status was only two groups, but originally Celiac was split in two, so there were three. An F-test tests if any of the pairwise tests are significant. Hui reported those p-values to you.  

#Secondly, we estimate the beta diversities of the different disease status at different age and sorting. We use the Bray-Curtis distance. The plots are at [beta_diversities_plot_1.pdf](./beta_diversities_plot_1.pdf), and [beta_diversities_plot_2.pdf](./beta_diversities_plot_2.pdf).The corresponding ANOVA test p-values are at Excel sheets [beta_diversities_plot_1_anova_pval.xlsx](./beta_diversities_plot_1_anova_pval.xlsx) and [beta_diversities_plot_2_anova_pval.xlsx](./beta_diversities_plot_2_anova_pval.xlsx).   




## Overview of relative abundance
#```{r ra}
levs <- c("Phylum", "Class", "Order", "Family", "Genus")
ra <- list()
pdf("relative_abundance_1.pdf", 12, 7)
for(lev in levs) {
  
  nps_lev <- nps %>%
    tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
    psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  ra[[lev]] <- nps_lev
  
  ggp <- ggplot(nps_lev, aes_string(x = "Disease.status", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid(Sort ~ Age) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()
writexl::write_xlsx(ra, "relative_abundance.xlsx")


##Statistics Version
levs <- c("Phylum", "Class", "Order", "Family", "Genus")
ra <- list()
stat_results <- list()

pdf("relative_abundance_1.pdf", 12, 7)
for(lev in levs) {
  
  nps_lev <- nps %>%
    tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
    psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                  # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                      # filter low abundant taxa
  ra[[lev]] <- nps_lev
  
  # Statistical testing
  test_result <- do.call(rbind, lapply(split(nps_lev, nps_lev$Disease.status), function(df) {
    t_test <- t.test(Abundance ~ Sort, data = df)
    data.frame(Level = lev, Group = df$Disease.status[1], p_value = t_test$p.value)
  }))
  stat_results[[lev]] <- test_result
  
  # Plotting
  ggp <- ggplot(nps_lev, aes_string(x = "Disease.status", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid(Sort ~ Age) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()


test_result <- do.call(rbind, lapply(split(nps_lev, nps_lev$Disease.status), function(df) {
  if (length(unique(df$Sort)) == 2) {
    t_test <- t.test(Abundance ~ Sort, data = df)
    return(data.frame(Level = lev, Group = df$Disease.status[1], p_value = t_test$p.value))
  } else {
    return(data.frame(Level = lev, Group = df$Disease.status[1], p_value = NA))
  }
}))

test_result <- do.call(rbind, lapply(split(nps_lev, nps_lev$Disease.status), function(df) {
  if (length(unique(df$Sort)) > 1) {
    # Replace with appropriate test (ANOVA or Kruskal-Wallis)
    test_res <- anova(lm(Abundance ~ Sort, data = df))
    p_value <- summary(test_res)[[1]]["Sort", "Pr(>F)"]
  } else {
    p_value <- NA
  }
  return(data.frame(Level = lev, Group = df$Disease.status[1], p_value = p_value))
}))




###Last Try
test_result <- do.call(rbind, lapply(split(nps_lev, nps_lev$Disease.status), function(df) {
  if (length(unique(df$Sort)) > 1) {
    # Replace with appropriate test (ANOVA or Kruskal-Wallis)
    test_res <- anova(lm(Abundance ~ Sort, data = df))
    p_value <- summary(test_res)[[1]]["Sort", "Pr(>F)"]
  } else {
    p_value <- NA
  }
  return(data.frame(Level = lev, Group = df$Disease.status[1], p_value = p_value))
}))


test_result <- do.call(rbind, lapply(split(nps_lev, nps_lev$Disease.status), function(df) {
  if (length(unique(df$Sort)) > 1) {
    test_res <- anova(lm(Abundance ~ Sort, data = df))
    p_value <- ifelse("Sort" %in% rownames(test_res), test_res["Sort", "Pr(>F)"], NA)
  } else {
    p_value <- NA
  }
  return(data.frame(Level = lev, Group = df$Disease.status[1], p_value = p_value))
}))




# Writing results
writexl::write_xlsx(ra, "relative_abundance.xlsx")
writexl::write_xlsx(stat_results, "relative_abundance_statistics.xlsx")


pdf("relative_abundance_2.pdf", 12, 7)
for(lev in levs) {
  
  nps_lev <- nps %>%
    tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
    psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  
  ggp <- ggplot(nps_lev, aes_string(x = "Age", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid(Sort ~ Disease.status) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()


#```

#We agglomerate the relative abundance (proportion) for each taxonomy levels, namely "Phylum", "Class", "Order", "Family", and "Genus". After that, we calculate the average relative abundance for each disease status, and then we filter out low abundance taxa (average relative abundance < 0.01). The plots for average relative abundance are at  [relative_abundance_1.pdf](./relative_abundance_1.pdf), and [relative_abundance_2.pdf](./relative_abundance_2.pdf)  

#```{r}
nps.presorted <- subset_samples(nps, Sorted == "Presort") 

levs <- c("Phylum", "Class", "Order", "Family", "Genus")
pdf("relative_abundance_3_presort.pdf", 12, 5)
for(lev in levs) {
  
  nps_lev <- nps.presorted %>%
    tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
    psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  
  ggp <- ggplot(nps_lev, aes_string(x = "Disease.status", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid( ~ Age) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()
#```

#```{r}
library(car)
set.seed(100)

otu <- matrix(otu_table(nps))
adjust <- min(otu[otu !=0 ])/2
ttest_res <- ftest_res <- list()
for(lev in levs) {
  
  nps_lev <- nps %>% 
    tax_glom(taxrank = lev) %>%
    transform_sample_counts(fun = function(x) logit(x, adjust = adjust)/log(2))
  
  for (s in levels(sample_data(nps_lev)$Sort)){
    for(a in levels(sample_data(nps_lev)$Age)) {
      nps_lev_tmp <- subset_samples(nps_lev, Sort== s & Age == a) 
      ttest_res_tmp <- mt(nps_lev_tmp, classlabel = "Disease.status", test = "t", B = 1e4)
      
      nps_lev_avg <- t(apply(otu_table(nps_lev_tmp), MARGIN = 2, FUN= function(v) tapply(v,
                                                                                         INDEX = sample_data(nps_lev_tmp)$Disease.status, FUN =  mean)))
      colnames(nps_lev_avg) <- paste0(colnames(nps_lev_avg), ".logit.avg")
      nps_lev_sem <- t(apply(otu_table(nps_lev_tmp), MARGIN = 2, FUN= function(v) tapply(v,
                                                                                         INDEX = sample_data(nps_lev_tmp)$Disease.status, 
                                                                                         FUN = function(x) sd(x)/sqrt(length(x)))))
      colnames(nps_lev_sem) <- paste0(colnames(nps_lev_sem), ".logit.sem")
      ttest_res[[paste(lev, s, a, sep = "_")]] <- data.frame(ttest_res_tmp, Sort = s, Age = a, 
                                                             nps_lev_avg[rownames(ttest_res_tmp), ],
                                                             nps_lev_sem[rownames(ttest_res_tmp), ])
    }
  }
  
  for (s in levels(sample_data(nps_lev)$Sort)){
    for(d in levels(sample_data(nps_lev)$Disease.status)){
      nps_lev_tmp <- subset_samples(nps_lev, Sort== s & Disease.status == d) 
      ftest_res_tmp <- mt(nps_lev_tmp, classlabel = "Age", test = "f", B = 1e4)
      
      nps_lev_avg <- t(apply(otu_table(nps_lev_tmp), MARGIN = 2, FUN= function(v) tapply(v,
                                                                                         INDEX = sample_data(nps_lev_tmp)$Age, FUN =  mean)))
      colnames(nps_lev_avg) <- paste0(colnames(nps_lev_avg), ".logit.avg")
      nps_lev_sem <- t(apply(otu_table(nps_lev_tmp), MARGIN = 2, FUN= function(v) tapply(v,
                                                                                         INDEX = sample_data(nps_lev_tmp)$Age, 
                                                                                         FUN = function(x) sd(x)/sqrt(length(x)))))
      colnames(nps_lev_sem) <- paste0(colnames(nps_lev_sem), ".logit.sem")
      ftest_res[[paste(lev, s, d, sep = "_")]] <- data.frame(ftest_res_tmp, Sort = s, Disease.status = d,
                                                             nps_lev_avg[rownames(ftest_res_tmp), ],
                                                             nps_lev_sem[rownames(ftest_res_tmp), ], 
                                                             check.names = FALSE)
    }
  }
}

ttest_res <- Reduce(rbind, ttest_res)
write.csv(df_signif(ttest_res, 3), "relative_abundance_1_ttest_disease_stats.csv")

ftest_res <- Reduce(rbind, ftest_res)
write.csv(df_signif(ftest_res, 3), "relative_abundance_2_ftest_age_stats.csv")
#```


## OTUs binary Venn diagrams
#```{r venn}
## Disease
res <- list()
pdf("sort_disease_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(sort in levels(md$Sort)){
  for(dis in levels(md$Disease.status)){
    md_tmp <- md[md$Sort == sort & md$Disease.status == dis, ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Age) > 0)
    nm <- paste(sort, "of", dis)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
    
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
  }
}
dev.off()

writexl::write_xlsx(res, "sort_disease_venn_tab.xlsx")

## Age
res <- list()
pdf("sort_age_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(sort in levels(md$Sort)){
  for(age in levels(md$Age)){
    md_tmp <- md[md$Sort == sort & md$Age == age, ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Disease.status) > 0)
    nm <- paste(sort, "of", age)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
    
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
  }
}
dev.off()

writexl::write_xlsx(res, "sort_age_venn_tab.xlsx")

## Sort
res <- list()
pdf("disease_age_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(dis in levels(md$Disease.status)){
  for(age in levels(md$Age)){
    md_tmp <- md[md$Disease.status == dis & md$Age == age & md$Sorted == "Postsort", ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Sort) > 0)
    nm <- paste(dis, "of", age)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
    
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
  }
}
dev.off()

writexl::write_xlsx(res, "disease_age_venn_tab.xlsx")
#```

#Venn diagrams are at [sort_disease_venn.pdf](./sort_disease_venn.pdf), [sort_age_venn.pdf](./sort_age_venn.pdf), and [disease_age_venn.pdf](./disease_age_venn.pdf). The corresponding Excel sheets are at [sort_disease_venn_tab.xlsx](./sort_disease_venn_tab.xlsx), [sort_age_venn_tab.xlsx](./sort_age_venn_tab.xlsx), and [disease_age_venn_tab.xlsx](./disease_age_venn_tab.xlsx).   

#```{r save}
write_phyloseq(ps,  type = "OTU")
write_phyloseq(ps,  type = "TAXONOMY")
write_phyloseq(ps,  type = "METADATA")
#```


### EDGER SCRIPT
#knitr::opts_chunk$set(echo = FALSE, include = FALSE)
#setwd("B:/")
source("fcns/config.r")
source("fcns/select_top.r")
source("fcns/select_top10.r")
source("fcns/select_top15.r")
source("fcns/fcns_limma/edger_contrasts.r")
#setwd("external/altindis/celiac_disease_16sn/edger")
library(plyr)
library(reshape2)
library(edgeR)
library(ggpubr)
#```

## Purpose
#Differential OTUs by using edger [1]
#Results are in folder [edger](../edger).

## Data
#OTU table [otu_table.csv](../phyloseq/otu_table.csv), 
#Taxonomy table [taxonomy_table.csv](../phyloseq/taxonomy_table.csv), 
#Metadata table [metadata_table.csv](../phyloseq/metadata_table.csv)

library(phyloseq)
library(writexl)

# Assuming 'ps' is your phyloseq object
otu_table_df <- as.data.frame(as.matrix(otu_table(ps)))
tax_table_df <- as.data.frame(as.matrix(tax_table(ps)))

# Convert 'sample_data' to a standard data frame and clean up additional attributes
metadata_df <- as.data.frame(sample_data(ps))
class(metadata_df) <- "data.frame"

# Writing to Excel files
write_xlsx(otu_table_df, "phyloseq_otu_table.xlsx")
write_xlsx(tax_table_df, "phyloseq_tax_table.xlsx")
write_xlsx(metadata_df, "phyloseq_metadata.xlsx")


counts <- otu_table_df

pheno <- metadata_df
annot <- tax_table_df

# Transpose the counts data frame
counts_transposed <- t(counts)

counts <- counts_transposed

# Updating the rownames if necessary
#rownames(counts_transposed) <- rownames(pheno)

rownames(counts) <- rownames(annot) <- paste0("ASV", 1:nrow(counts))






#DD changed "OTU" to "ASV"
#rownames(counts) <- rownames(annot) <- paste0("OTU", 1:nrow(counts))
rownames(counts) <- rownames(annot) <- paste0("ASV", 1:nrow(counts))

##DD ADDITION FOR ICI PART 10.14.22
library(dplyr)
library(tibble)
composite <- counts
composite <- as.data.frame(composite)
counts <- as.data.frame(counts)

composite <- composite %>%
  rownames_to_column(var="ASV")


deneme<- row.names(composite) 
deneme <- mutate(composite, ASV=row.names(composite))

deneme <- cbind(rownames(counts),rownames(counts_yedek))

asv_sequences <-  deneme
#asv_sequences <- as.data.frame(asv_sequences)
#setnames(asv_sequences, c("ASV","OTU"))

#annot$Taxa <- apply(annot, 1, function(v) paste(v[!is.na(v)], collapse = "_"))
annot_species <- annot %>% filter(!is.na(Species))
annot_other <- annot %>% filter(is.na(Species))

annot_species$Taxa <- apply(annot_species,1, function(v) paste((tail(v[!is.na(v)],2) ),collapse = " "))
annot_other$Taxa <- apply(annot_other,1, function(v) paste((tail(v[!is.na(v)],1) ),collapse = " "))

annot_combined <- rbind(annot_species,annot_other)

annot <- annot_combined

#annot <- annot_species

#annot$Taxa <- apply(annot,1, function(v) paste((tail(v[!is.na(v)],2) ),collapse = " "))
pheno$Disease.status <- factor(pheno$Disease.status, levels = c("Control", "Celiac"))
pheno$Group <- factor(pheno$Group, levels = unique(pheno$Group))
```

#species_hit <- annot %>% filter(!is.na(Species))
#deneme <- annot %>% filter(!is.na(Species))
#annot$Taxanew <- apply(annot, 1, function(v) paste(v[!is.na(v)], collapse = " "))
#annot$deneme <- apply(annot, 1, function(v) paste(tail(v[!is.na(v)],2)),collapse = " ")

#annot$deneme123 <- apply(annot,1, function(v) (tail(v[!is.na(v)],2) )
#annot$deneme123 <- apply(annot,1, function(v) paste((tail(v[!is.na(v)],2) ),collapse = " "))

We combine the counts of each two technical replicates.   

## Filtering
```{r filt}
dge <- DGEList(counts = counts)
dge <- dge[rowSums(counts >= 10) > 5, ]
dim(dge)

# donot using  calcNormFactors()
```

To filter out low abundant OTUs, we keep OTUs that have read counts at least 10 in 5 samples. There are 611 OTUs after filtering. 

## Test for differential OUTs
```{r edger}
## Define contrasts for age 1
contr.v <- c(Celiac_vs_Control_in_Presort_Age1 = "Presort_Celiac_1 - Presort_Control_1",
             Celiac_vs_Control_in_IGneg_Age1 = "IGneg_Celiac_1 - IGneg_Control_1",
             Celiac_vs_Control_in_IGpos_Age1 = "IGpos_Celiac_1 - IGpos_Control_1", 
             IGpos_vs_IGneg_in_Control_Age1 = "IGpos_Control_1 - IGneg_Control_1",  
             IGpos_vs_IGneg_in_Celiac_Age1 = "IGpos_Celiac_1 - IGneg_Celiac_1")

pdf("edger_qc_metrics.pdf")
res <- edger.contrasts(dge, grp = pheno$Group, contrasts.v = contr.v, plot = TRUE)
dev.off()
mtt <- res$mtt; logcpm <- res$logcpm
rm(res)
mtt.df <- data.frame(signif(mtt, 3), annot[rownames(mtt), ])
source("fcns/fcns_plot/sig_hist.r")
sig.hist(mtt, name = "signif_hist_edgeR", pi0 = TRUE)
write.csv(mtt.df, "otu_stats_edgeR.csv", na = "")
#```

#To discover the differential OTUs, we use edgeR, an R package for differential expression analysis of digital gene expression data [1]. We perform empirical Bayes quasi-likelihood F-tests for the following comparisons: between any 2 disease status at different age and sorting, or between IGpos and IGneg in different age and disease status.    
#The histograms of significance are at [signif_hist.pdf](./signif_hist.pdf). If no OTUs were associated with the phenotype, we would expect the p-value histogram to be flat and all FDRs to be near one. The more associated OTUs there are, the more enrichment there is at low p-values, the lower will be the FDRs.  We also estimate the proportion of the true null hypothesis (i.e. non-significant OTUs) [2].    
#OTU statistics tables for all OTUs [otu_stats.csv](./otu_stats.csv). The table contains the average logCPM of each group, p-values, FDR, log fold-change, fold-change, and taxonomy information.      

## Plots
#```{r plots}
## boxplot
topotu1 <- select.top(mtt, contrasts.v = contr.v[grep("Presort", names(contr.v))])
pdf("top_otus_boxplot_presort.pdf", 9, 3)
for (otu in topotu1){
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
  ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test")
  plot(ggp)
}
dev.off()


#####
source("fcns/select_top.r")
pdf("top_otus_violin_plots_presort.pdf", 9, 3)
for (otu in topotu1){
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
  #dat2p <- data.frame(logCPM = logcpm[otu, ], pheno)
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM, fill = Disease.status))+ theme_bw()
  #ggp <- ggp + ggtitle(annot[otu, "Taxa"]) + theme(legend.position = "none")
  
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none") ## DD ADDITION
  
  ggp <- ggp + geom_violin(trim = FALSE, alpha = 0.3) + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)
  ggp <- ggp + stat_summary(fun.data = "mean_cl_boot", geom = "crossbar", colour = "black", width = 0.1) ##DD addition for median bar
  ggp <- ggp + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

#####

#topotu2 <- select.top(mtt, contrasts.v = contr.v[-grep("Presort", names(contr.v))])
topotu2 <- select.top(mtt, contrasts.v = contr.v)
pdf("top_otus_boxplot_postsort.pdf", 9, 5)
for (otu in topotu2){
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
  ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(Sort~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test")
  plot(ggp)
}
dev.off()

#####
source("fcns/select_top.r")
pdf("top_otus_violin_plots_postsort.pdf", 9, 3)
for (otu in topotu2){
  #dat2p <- data.frame(logCPM = logcpm[otu, ], pheno)
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM, fill = Disease.status)) + theme_bw()
  #ggp <- ggp + ggtitle(annot[otu, "Taxa"]) + theme(legend.position = "none")
  
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none") ## DD ADDITION
  
  ggp <- ggp + geom_violin(trim = FALSE, alpha = 0.3) + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)
  
  #ggp <- ggp +  stat_summary(fun.data = "mean_sdl")
  ggp <- ggp + stat_summary(fun.data = "mean_cl_boot", geom = "crossbar", colour = "black", width = 0.1)
  ggp <- ggp + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

#####
BiocManager::install("ggstatsplot",force = TRUE)
library(ggstatsplot)
pdf("top_otus_violin_plots_postsort15.pdf", 9, 3)
for (otu in topotu2){
  #dat2p <- data.frame(logCPM = logcpm[otu, ], pheno)
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM, fill = Disease.status)) + theme_bw()
  
  #ggp <- ggbetweenstats(data = dat2p, x = Disease.status, y = logCPM)
  #ggp <- ggp + ggtitle(annot[otu, "Taxa"]) + theme(legend.position = "none")
  
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none") ## DD ADDITION
  
  ggp <- ggp + geom_violin(trim = FALSE, alpha = 0.3) + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)+
    stat_summary(fun.data = "mean_cl_boot", geom = "crossbar",
                 colour = "black", width = 0.1)+stat_compare_means(method = "t.test")
  #ggo <- ggp + stat_summary(fun.data="mean_sdl", mult=1, geom="crossbar", width=0.2 )
  ggp <- ggp + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)
  
  #ggp <- ggp +  stat_summary(fun.data = "mean_sdl")
  ggp <- ggp + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

## boxplot for all
pdf("all_otus_boxplot_presort.pdf", 9, 3)
for (otu in rownames(logcpm)){
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
  ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test")
  plot(ggp)
}
dev.off()

#####
source("fcns/select_top.r")
pdf("all_otus_violin_plots_presort.pdf", 9, 3)
for (otu in  rownames(logcpm)){
  #dat2p <- data.frame(logCPM = logcpm[otu, ], pheno)
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM, fill = Disease.status)) + theme_bw()
  #ggp <- ggp + ggtitle(annot[otu, "Taxa"]) + theme(legend.position = "none")
  
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none") ## DD ADDITION
  
  ggp <- ggp + geom_violin(trim = FALSE, alpha = 0.3) + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)
  ggp <- ggp + stat_summary(fun.data = "mean_cl_boot", geom = "crossbar", colour = "black", width = 0.1) ##DD addition
  ggp <- ggp + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

#####

pdf("all_otus_boxplot_postsort.pdf", 9, 5)
for (otu in rownames(logcpm)){
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
  ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(Sort~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test")
  plot(ggp)
}
dev.off()

#####
source("fcns/select_top.r")
pdf("all_otus_violin_plots_postsort.pdf", 9, 3)
for (otu in  rownames(logcpm)){
  #dat2p <- data.frame(logCPM = logcpm[otu, ], pheno)
  dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
  
  ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM, fill = Disease.status)) + theme_bw()
  #ggp <- ggp + ggtitle(annot[otu, "Taxa"]) + theme(legend.position = "none")
  
  ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none") ## DD ADDITION
  
  ggp <- ggp + geom_violin(trim = FALSE, alpha = 0.3) + geom_jitter(shape = 21, position = position_jitter(0.2), alpha = 0.5)
  ggp <- ggp + stat_summary(fun.data = "mean_cl_boot", geom = "crossbar", colour = "black", width = 0.1) ## DD ADDITION
  ggp <- ggp + facet_grid(~ Age)+scale_fill_manual(values=c( "#00BFC4","#F8766D"))
  #stat_compare_means(label.x=2,label.y=0.5, method = "t.test") 
  plot(ggp)
}
dev.off()

#####

## Clostridium XlVa
otus <- readLines("Clostridium_XlVa.txt")
samp <- rownames(pheno)[pheno$Age == "5year"]
dat2p <- cbind(t(logcpm[otus, samp]), pheno[samp, c("Disease.status","Sort")])
dat2p <- melt(dat2p, id.vars =  c("Disease.status","Sort"),  variable.name = "OTU",value.name = "logCPM")

pdf("Clostridium_XlVa_boxplot.pdf", 6, 12)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle("Age: 5 years, Clostridium XlVa") + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(OTU~Sort)
plot(ggp)
dev.off()

####pool 6
otus <- readLines("Clostridium_XlVa.txt")
samp <- rownames(pheno)[pheno$Age == "5year" & pheno$Sort == "Presort"]
dat2p <- cbind(logCPM = colMeans(logcpm[otus, samp]), pheno[samp, "Disease.status", drop = FALSE])
pval <- signif(t.test(logCPM~Disease.status, data = dat2p)$p.value, 3)

pdf("Clostridium_XlVa_boxplot2_presort.pdf", 3.5, 3.5)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle(paste("Age: 5 years, Presort, \n6 Clostridium XlVa, pval =", pval)) + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status))
plot(ggp)
dev.off()

###pool all
otus <- rownames(mtt.df)[which(mtt.df$Genus == "Clostridium_XlVa")]
samp <- rownames(pheno)[pheno$Age == "5year" & pheno$Sort == "Presort"]
dat2p <- cbind(logCPM = colMeans(logcpm[otus, samp]), pheno[samp, "Disease.status", drop = FALSE])
pval <- signif(t.test(logCPM~Disease.status, data = dat2p)$p.value, 3)

pdf("Clostridium_XlVa_boxplot3_presort.pdf", 3.5, 3.5)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle(paste("Age: 5 years, Presort, \nAll Clostridium XlVa, pval =", pval)) + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status))
plot(ggp)
dev.off()

source("fcns/ezheat.R")
source("fcns/prune_mat.R")
## heatmaps
pheno.df1 <- pheno[pheno$Sort == "Presort", c("Disease.status", "Age")]
gaps.col1 <- which(diff(as.numeric(pheno.df1$Age), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Sort == "Presort"], pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col1, main = "logCPM", name = "top_otus_heat_presort", height = 8, width = 16, clip = 2)


source("fcns/ezheat.R")
source("fcns/prune_mat.R")
## heatmaps

source("fcns/ezheat.R")
source("fcns/prune_mat.R")

# Assuming pheno, logcpm, topotu1, and annot are correctly defined
pheno.df1 <- pheno[pheno$Sort == "Presort", "Disease.status"]

# Check the lengths
if(length(topotu1) != length(annot[topotu1, "Taxa"])) {
  stop("Mismatch in number of top OTUs and their corresponding taxonomic labels")
}

# Prepare the data for heatmap
heatmap_data <- logcpm[topotu1, pheno$Sort == "Presort"]

# Check if dimensions match
if(nrow(heatmap_data) != length(topotu1)) {
  stop("Mismatch in dimensions of heatmap data and top OTUs")
}

# Create the heatmap
ezheat(heatmap_data, pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)


source("fcns/ezheat.R")
source("fcns/prune_mat.R")

# Assuming pheno is your metadata dataframe and contains Disease.status and Sort columns
pheno.df1 <- pheno[pheno$Sort == "Presort", "Disease.status"]

# Log transform and subset the top OTUs for the Presort category
# Replace 'logcpm' and 'topotu1' with your actual variables/log-transformed counts and top OTUs' indices
ezheat(logcpm[topotu1, pheno$Sort == "Presort"], pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)



pheno.df1 <- pheno[pheno$Sort == "Presort", c("Disease.status", "Age")]
gaps.col1 <- which(diff(as.numeric(pheno.df1$Age), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Sort == "Presort"], pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col1, main = "logCPM", name = "top_otus_heat_presort", height = 8, width = 16, clip = 2)


pheno.df2 <- pheno[pheno$Sort != "Presort", c("Disease.status", "Age", "Sort")]
gaps.col2 <- which(diff(as.data.frame.numeric_version(pheno.df2$Age), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"], pheno.df = pheno.df2,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2], labcols = "",
       gaps_col = c(16,31,47,62,75,84,96), main = "logCPM", name = "top_otus_heat_postsort(10.3.22)", height = 8, width = 25, clip = 2)

pheno.df3 <- pheno[pheno$Sort != "Presort" & pheno$Age== "5year", c("Disease.status", "Sort")]
gaps.col3 <- which(diff(as.numeric(pheno.df3$Age), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"& pheno$Age== "5year"], pheno.df = pheno.df3,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col3, main = "logCPM", name = "top_otus_heat_postsort_5year(9.29.22)", height = 8, width = 12.5, clip = 2)

pheno.df4 <- pheno[pheno$Sort != "Presort" & pheno$Age== "2.5year", c("Disease.status", "Sort")]
gaps.col4 <- which(diff(as.numeric(pheno.df4$Age), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"& pheno$Age== "2.5year"], pheno.df = pheno.df4,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col4, main = "logCPM", name = "top_otus_heat_postsort_2_5year(7.8.22)", height = 8, width = 12.5, clip = 2)

pheno.df5 <- pheno[pheno$Sort == "Presort", c("Disease.status","Age")]
#gaps.col5 <- which(diff(as.numeric(pheno.df5$Age), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Age== "2.5year"], pheno.df = pheno.df5,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col5, main = "logCPM", name = "top_otus_heat_presort_2_5_year(7.8.22)", height = 8, width = 16, clip = 2)

pheno.df6 <- pheno[pheno$Sort == "Presort"& pheno$Age== "5year", c("Disease.status")]
gaps.col6 <- which(diff(as.numeric(pheno.df6$Age), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Sort == "Presort"& pheno$Age== "5year"], pheno.df = pheno.df6,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col6, main = "logCPM", name = "top_otus_heat_presort_5_year(7.8.22)", height = 8, width = 16, clip = 2)
```


###DD NEW HEATMAPS!
topotu7 <- select.top15(mtt, contrasts.v = contr.v[grep("Presort", names(contr.v))]) ###top30 otu presort

pheno.df7 <- pheno[pheno$Sort == "Presort", "Disease.status"]
gaps.col7 <- which(diff(as.numeric(pheno.df7$Age), lag=1) != 0)
ezheat(logcpm[topotu7, pheno$Sort == "Presort"], pheno.df = pheno.df7,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu7, "Taxa"], labcols = "",
       gaps_col = gaps.col7, main = "logCPM", name = "top15_otus_heat_presort", height = 8, width = 16, clip = 2)

Equal numbers of top OTUs (based on p-values) are selected from each comparison. The boxplots for top OUTs are at [top_otus_boxplot_presort.pdf](./top_otus_boxplot_presort.pdf), and [top_otus_boxplot_postsort.pdf](./top_otus_boxplot_postsort.pdf). The same sets of top OTUs are use in heatmaps [top_otus_heat_presort.pdf](./top_otus_heat_presort.pdf) and [top_otus_heat_postsort.pdf](./top_otus_heat_postsort.pdf).


###DD HEATMAPS AFTER REVIEW 10.11.22
## heatmaps
pheno.df1 <- pheno[pheno$Sort == "Presort", c("Disease.status")]
gaps.col1 <- which(diff(as.numeric(pheno.df1$Disease.status), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Sor t == "Presort"], pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col1, main = "logCPM", name = "top_otus_heat_presort", height = 8, width = 16, clip = 2)


pheno.df2 <- pheno[pheno$Sort != "Presort", c("Disease.status", "Age", "Sort")]
gaps.col2 <- which(diff(as.numeric(pheno.df1$Disease.status), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"], pheno.df = pheno.df2,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col2, main = "logCPM", name = "top_otus_heat_postsort(10.11.22)", height = 8, width = 25, clip = 2)

pheno.df3 <- pheno[pheno$Sort == "Presort" & pheno$Age== "1year", c("Disease.status", "Sort")]
gaps.col3 <- which(diff(as.numeric(pheno.df3$Disease.status), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"& pheno$Age== "1year"], pheno.df = pheno.df3,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col3, main = "logCPM", name = "top_otus_heat_postsort_5year(10.11.22)", height = 8, width = 12.5, clip = 2)

pheno.df4 <- pheno[pheno$Sort != "Presort" & pheno$Age== "2.5year", c("Disease.status", "Sort")]
gaps.col4 <- which(diff(as.numeric(pheno.df4$Disease.status), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"& pheno$Age== "2.5year"], pheno.df = pheno.df4,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col4, main = "logCPM", name = "top_otus_heat_postsort_2_5year(10.11.22)", height = 8, width = 12.5, clip = 2)

```{r check}
stopifnot(rownames(counts) == rownames(annot))
stopifnot(colnames(counts) == rownames(pheno))
stopifnot(mtt.df[1, "Taxa"] == "Bacteria_Proteobacteria_Gammaproteobacteria_Enterobacteriales_Enterobacteriaceae_Citrobacter")
```

source("fcns/ezheat.R")
source("fcns/prune_mat.R")

# Assuming pheno, logcpm, topotu1, and annot are correctly defined
pheno.df1 <- pheno[pheno$Sort == "Presort", "Disease.status"]

# Inspect the length and contents of topotu1 and corresponding labels in annot
print(length(topotu1))
print(sum(topotu1 %in% rownames(annot)))  # This should be equal to length(topotu1)

# Update topotu1 or annot as necessary based on the inspection results

# Prepare the data for heatmap
heatmap_data <- logcpm[topotu1, pheno$Sort == "Presort"]

# Ensure the dimensions match
if(nrow(heatmap_data) != length(topotu1)) {
  stop("Mismatch in dimensions of heatmap data and top OTUs")
}

# Create the heatmap
ezheat(heatmap_data, pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)


# Verify the number of OTUs and labels
if (nrow(heatmap_data) != length(annot[topotu1, "Taxa"])) {
  stop("Number of rows in heatmap data does not match the number of labels in labrows")
}

# Adjust topotu1 if necessary
# For example, ensure that topotu1 only includes OTUs with labels in annot
# topotu1 <- topotu1[which(rownames(annot)[topotu1] %in% rownames(heatmap_data))]

# Create the heatmap with verified data
ezheat(heatmap_data, pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)


# Filter topotu1 to include only indices with labels in annot
valid_indices <- which(rownames(annot) %in% rownames(heatmap_data))
topotu1_filtered <- topotu1[topotu1 %in% valid_indices]

# Recreate heatmap data with the filtered topotu1
heatmap_data_filtered <- logcpm[topotu1_filtered, pheno$Sort == "Presort"]

# Verify the number of OTUs and labels again
if (nrow(heatmap_data_filtered) != length(annot[topotu1_filtered, "Taxa"])) {
  stop("Number of rows in filtered heatmap data does not match the number of labels in labrows")
}

# Create the heatmap with verified data
ezheat(heatmap_data_filtered, pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1_filtered, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)


# Filter topotu1 to include only indices with labels in annot
valid_indices <- which(rownames(annot) %in% rownames(heatmap_data))
topotu1_filtered <- topotu1[topotu1 %in% valid_indices]

# Check the length of topotu1_filtered
if(length(topotu1_filtered) < 2) {
  stop("Not enough OTUs after filtering. Need at least 2 OTUs for heatmap clustering.")
}

# Recreate heatmap data with the filtered topotu1
heatmap_data_filtered <- logcpm[topotu1_filtered, pheno$Sort == "Presort"]

# Verify the number of OTUs and labels again
if (nrow(heatmap_data_filtered) != length(annot[topotu1_filtered, "Taxa"])) {
  stop("Number of rows in filtered heatmap data does not match the number of labels in labrows")
}

# Create the heatmap with verified data
ezheat(heatmap_data_filtered, pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = annot[topotu1_filtered, "Taxa"], labcols = "", main = "logCPM", name = "top_otus_heat_presort", 
       height = 8, width = 16, clip = 2)


###PICRUST RMD

---
title: "Function Prediction"
author: "Hui/Jonathan"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: word_document
bibliography: B:/annotations/bib/bioinfo.bib
---

```{bash, echo = FALSE, eval = FALSE}
ssh hp135@o2.hms.harvard.edu 
cd /n/data1/joslin/cores/bbcore/celiac_disease_16sn/picrust2
srun -c 4 -p interactive --pty --mem-per-cpu=8G -t 0-12 /bin/bash
module load gcc/6.2.0 R/3.5.1; R
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = FALSE)
#setwd("/home/hp135")
setwd("B:/")
source("fcns/config.r")
source("fcns/fcns_limma/edger_contrasts.r")
#setwd("/n/data1/joslin/cores/bbcore/celiac_disease_16sn/picrust2")
setwd("external/altindis/celiac_disease_16sn/picrust2")
library(biomformat)
library(dada2)
```

## Data
OTU table [seqtab_nochim.rds](../dada2/seqtab_nochim.rds), 
Metadata table [taxonomy_table.csv](../phyloseq/metadata_table.csv), 

## Make biom and fasta file
```{r biom}
seqtab_nochim <- readRDS("../dada2/seqtab_nochim.rds")
uniquesToFasta(seqtab_nochim, "seq_nochim.fa", ids = paste0("Seq", 1:ncol(seqtab_nochim)))

seqtab_nochim_matched <- seqtab_nochim
colnames(seqtab_nochim_matched) <- paste0("Seq", 1:ncol(seqtab_nochim))
st.biom <- make_biom(t(seqtab_nochim_matched))
write_biom(st.biom, "seq_nochim.biom")
```

```{r}
library(seqinr)
seqs <- read.fasta("seq_nochim.fa")
seqs <- sapply(seqs, paste, collapse = "")
seqs <- toupper(as.vector(seqs))
stopifnot(seqs == colnames(seqtab_nochim))
```

## Install PICRUSt2
```{bash, eval=FALSE}
q()
n
# create a new conda environment
cd ~
module purge
module load conda2/4.2.13
conda create -n picrust2 -c bioconda -c conda-forge picrust2=2.2.0_b
```

## Run PICRUSt2
```{bash, eval=FALSE}
# activate the environment  
source activate picrust2
cd /n/data1/joslin/cores/bbcore/celiac_disease_16sn/picrust2
picrust2_pipeline.py -s seq_nochim.fa -i seq_nochim.biom -o picrust2_out_pipeline -p 4
# deactivate virtual environment
conda deactivate
```

We perform function prediction by PICRUSt2.   

```{r parse}
# meta data
md <- read.csv("../data/CeD new label (2).csv")
md_age <- read.csv("../data/Celiakibarn 190906.csv")
md <- merge(md, md_age, all = TRUE)
rm(md_age)

md <- md[order(md$New.Sample.ID), ]
md <- md[rep(1:nrow(md), each = 3),]
md$New.Sample.ID <- paste0("S", md$New.Sample.ID, c("", ".neg", ".pos"))
md$Sort <- rep(c("Presort", "IGneg", "IGpos"), time = nrow(md)/3)
md$Disease.status <- sapply(md$Disease.status, FUN = switch, A = "Celiac", B = "Control")
md$ABIS.no <- paste0("ABIS", md$ABIS.no)

md$Disease.status <- factor(md$Disease.status, levels = c("Control", "Celiac"))
md$Sorted <- md$Sort <- factor(md$Sort, levels = c("Presort", "IGneg", "IGpos"))
levels(md$Sorted) <- list(Presort = "Presort", Postsort = c("IGneg", "IGpos"))
md <- md[order(md$Sort, md$Disease.status, md$Age, md$ABIS.no), ]
md$Group <- paste(md$Sort, md$Disease.status, md$Age, sep = "_")
md$Group <- factor(md$Group, levels = unique(md$Group))
md$Age <- factor(paste0(md$Age, "year"))
md <- data.frame(md, row.names = "New.Sample.ID")

## pwys data
pwys.tab <- read.delim("picrust2_out_pipeline/pathways_out/path_abun_unstrat.tsv.gz", row.names = 1)
# dim(pwys.tab)
grp <- gsub("\\.[1-2]$", "", gsub("\\.[1-2]\\.", ".", colnames(pwys.tab)))
pwys.tab <- t(rowsum(t(pwys.tab), group = grp))
rm(grp)
# dim(pwys.tab)
nms <- intersect(rownames(md), colnames(pwys.tab))
pwys.tab <- pwys.tab[,nms]
md <- md[nms, ]

## named pwys
pwys.nm <- setNames(rownames(pwys.tab), nm=rownames(pwys.tab))
pwys.nm <- gsub("PWY(|0)-[0-9]+$|P[0-9]+$", NA,
                gsub("-PWY$", "", pwys.nm))
```

## Annotate pathways
```{r annot}
annot <- read.delim("B:/annotations/metacyc/All_pathways_of_MetaCyc.txt", row.names = 1)
colnames(annot) <- "Pathway"
annot$Pathway <- gsub("<[A-Za-z]>|</[A-Za-z]>", "", annot$Pathway)
```

## Filter and log transformation
```{r filt}
pwys.tab <- pwys.tab[rowSums(pwys.tab >= 10) > 5, ]
# range(pwys.tab)
pwys.tab <- log2(pwys.tab+1)


boxplot(pwys.tab)
hist(pwys.tab)

v <- voom(2^pwys.tab, plot = TRUE)

#save(pwys.tab, md, annot, file = "pwys.rda")
```

To filter out low abundant pathways, we keep KO pathways that have abundance at least 10 in 5 samples. There are 377 KO terms after filtering. We then perform log transfomation on the pathway abundance.  

## Test for differential KO terms
```{r limma}
contr.v <- c(Celiac_vs_Control_in_Presort_Age1 = "Presort_Celiac_1 - Presort_Control_1",
             Celiac_vs_Control_in_IGneg_Age1 = "IGneg_Celiac_1 - IGneg_Control_1",
             Celiac_vs_Control_in_IGpos_Age1 = "IGpos_Celiac_1 - IGpos_Control_1", 
             IGpos_vs_IGneg_in_Control_Age1 = "IGpos_Control_1 - IGneg_Control_1",  
             IGpos_vs_IGneg_in_Celiac_Age1 = "IGpos_Celiac_1 - IGneg_Celiac_1",
             
             Celiac_vs_Control_in_Presort_Age2.5 = "Presort_Celiac_2.5 - Presort_Control_2.5",
             Celiac_vs_Control_in_IGneg_Age2.5 = "IGneg_Celiac_2.5 - IGneg_Control_2.5",
             Celiac_vs_Control_in_IGpos_Age2.5 = "IGpos_Celiac_2.5 - IGpos_Control_2.5", 
             IGpos_vs_IGneg_in_Control_Age2.5 = "IGpos_Control_2.5 - IGneg_Control_2.5",  
             IGpos_vs_IGneg_in_Celiac_Age2.5 = "IGpos_Celiac_2.5 - IGneg_Celiac_2.5",
             
             Celiac_vs_Control_in_Presort_Age5 = "Presort_Celiac_5 - Presort_Control_5",
             Celiac_vs_Control_in_IGneg_Age5 = "IGneg_Celiac_5 - IGneg_Control_5",
             Celiac_vs_Control_in_IGpos_Age5 = "IGpos_Celiac_5 - IGpos_Control_5", 
             IGpos_vs_IGneg_in_Control_Age5 = "IGpos_Control_5 - IGneg_Control_5",  
             IGpos_vs_IGneg_in_Celiac_Age5 = "IGpos_Celiac_5 - IGneg_Celiac_5")


mtt <- limma_contrasts(pwys.tab, grp = md$Group, contrast.v = contr.v, trend = TRUE)
mtt.df <- data.frame(signif(mtt, 3), annot[rownames(mtt), , drop = FALSE])

signif_hist(mtt, name = "signif_hist", pi0 = TRUE)
write.csv(mtt.df, "pwys_stats.csv", na = "")
```

To discover the differential KO terms, we use limma, an R package that powers differential expression analyses [@ritchie_2015]. We perform We perform moderated t-test for the following comparisons: between any 2 disease status at different age and sorting, or between IGpos and IGneg in different age and disease status.    
The histograms of significance are at [signif_hist.pdf](./signif_hist.pdf). If no KO terms were associated with the phenotype, we would expect the p-value histogram to be flat and all FDRs to be near one. The more associated KO terms there are, the more enrichment there is at low p-values, the lower will be the FDRs.  We also estimate the proportion of the true null hypothesis (i.e. non-significant KO terms) [@langaas_2005].    
KO term statistics tables for all KO terms [pwys_stats.csv](./pwys_stats.csv). The table contains the average log2 abundance of each group, p-values, FDR, log fold-change, fold-change, and pathway information.      

## Plots
```{r plots}
## boxplot
top_pwy1 <- select.top(mtt.df, contrasts.v = contr.v[grep("Presort", names(contr.v))])
pdf("top_pwys_boxplot_presort.pdf", 9, 3)
for (pwy in top_pwy1){
    dat2p <- data.frame(logabun = pwys.tab[pwy, md$Sort == "Presort"], md[md$Sort == "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logabun)) + theme_bw()
    ggp <- ggp + ggtitle(substr(annot[pwy, "Pathway"], 1, 80)) + theme(legend.position = "none") + ylab("Log2 abundance")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(~ Age)
    plot(ggp)
}
dev.off()

top_pwy2 <- select.top(mtt.df, contrasts.v = contr.v[-grep("Presort", names(contr.v))])
pdf("top_pwys_boxplot_postsort.pdf", 9, 5)
for (pwy in top_pwy2){
    dat2p <- data.frame(logabun = pwys.tab[pwy, md$Sort != "Presort"], md[md$Sort != "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logabun)) + theme_bw()
    ggp <- ggp + ggtitle(substr(annot[pwy, "Pathway"], 1, 80)) + theme(legend.position = "none") + ylab("Log2 abundance")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(Sort~ Age)
    plot(ggp)
}
dev.off()

## heatmaps
md.df1 <- md[md$Sort == "Presort", c("Disease.status", "Age")]
gaps.col1 <- which(diff(as.numeric(as.factor(md.df1$Age)), lag=1) != 0)
ezheat(pwys.tab[top_pwy1, md$Sort == "Presort"], pheno.df = md.df1,  sc = "z", reorder_rows = TRUE, 
       labrows = substr(annot[top_pwy1, "Pathway"], 1, 80), labcols = "",
       gaps_col = gaps.col1, main = "Log2 abundance", name = "top_pwys_heat_presort", height = 8, width = 16, clip = 2)

md.df2 <- md[md$Sort != "Presort", c("Disease.status", "Age", "Sort")]
gaps.col2 <- which(diff(as.numeric(as.factor(md.df2$Age)), lag=1) != 0)
ezheat(pwys.tab[top_pwy2, md$Sort != "Presort"], pheno.df = md.df2,  sc = "z", reorder_rows = TRUE, 
       labrows = substr(annot[top_pwy2, "Pathway"], 1, 80), labcols = "",
       gaps_col = gaps.col2, main = "Log2 abundance", name = "top_pwys_heat_postsort", height = 8, width = 30, clip = 2)
```

Equal numbers of top KO terms (based on p-values) are selected from each comparison. The boxplots for top KO terms are at [top_pwys_boxplot_presort.pdf](./top_pwys_boxplot_presort.pdf), and [top_pwys_boxplot_postsort.pdf](./top_pwys_boxplot_postsort.pdf). The same sets of top KO terms are use in heatmaps [top_pwys_heat_presort.pdf](./top_pwys_heat_presort.pdf) and [top_pwys_heat_postsort.pdf](./top_pwys_heat_postsort.pdf).


```{r check}
stopifnot(rownames(pwys.tab) %in% rownames(annot))
stopifnot(colnames(pwys.tab) == rownames(md))
stopifnot(rownames(mtt.df)[1] == "PWY-7376")
stopifnot(!is.na(mtt.df$Pathway))
stopifnot(mtt.df$Pathway != "")
```




